### ISSUE TITLE
unique pointer to mpi solver

### ISSUE THREAD

**Issue by dboaretti** [ddemidov/amgcl#100]
Hi Denis,

it's still me, Daniele, with another issue. 
I wanted to use a unique pointer in my class for the MPI solver I define in one of the function of my custom class. 
The solver is of this type: 
```
    typedef amgcl::mpi::make_solver<
        amgcl::mpi::amg<
            Backend,
            amgcl::runtime::mpi::coarsening::wrapper<Backend>,
            amgcl::runtime::mpi::relaxation::wrapper<Backend>,
            amgcl::runtime::mpi::direct::solver<double>,
            amgcl::runtime::mpi::partition::wrapper<Backend>>,
        amgcl::runtime::solver::wrapper>
        Solver;
```
However, when I run the code with two MPI nodes, I see that it reaches the end but it ends up with this error, which should be an heap corruption error:

```
job aborted:
[ranks] message

[0-1] process exited without calling finalize

---- error analysis -----

[0-1] on LBB123-16
python ended prematurely and may have crashed. exit code 0xc0000374
```

However, in one of the old version of the code I use the following type of solver:
```
    typedef
    amgcl::make_solver<
        amgcl::amg<
            amgcl::backend::builtin<double>,
            amgcl::coarsening::smoothed_aggregation,
            amgcl::relaxation::spai0
            >,
        amgcl::solver::cg<amgcl::backend::builtin<double>>
    > Solver;
```
 and in the same way I use a unique pointer to the Solver I define later. In this case, the code ends without any error. 

I am wondering about the reason of the heap corruption error. Do you have any ideas?
If there is nothing to be added in the amgcl source code, what should I do to use the MPI solver defined in one function of my class in another function of the same class? I think that the pointer is the only way to deal with this. 
Let me know if you need further info. 
Thank you for your time.

**Comment by ddemidov** [comment]
Hi Daniele,

Do you call `MPI_Finalize` in your code? The error message usually occurs when you forget to. In my examples I am using `BOOST_SCOPE_EXIT` to call `MPI_Finalize` automatically at the end of the program:

https://github.com/ddemidov/amgcl/blob/47a2906a92ad011b6e9045be4778eae9ff1e95ae/examples/mpi/mpi_amg.cpp#L420-L422

**Comment by dboaretti** [comment]
I don't call it because I use `boost::mpi::environment env` and `boost::mpi::communicator world` in my class which already deal with this, for example like this https://www.boost.org/doc/libs/1_61_0/doc/html/mpi/tutorial.html . My code was able to run in parallel with at least 2 MPI nodes as well as as a serial program with these two boost objects.
In the attempt of integrating my code with this MPI solver I use the amgcl communicator to define the MPI solver, which I can define without errors at the end of program, but I encounter errors descried above when, in addition, I define the unique pointer to the solver defined just before. 
However, I can use 
https://github.com/ddemidov/amgcl/blob/47a2906a92ad011b6e9045be4778eae9ff1e95ae/examples/mpi/mpi_amg.cpp#L420-L422
at the end of the program as a workaround, but this works only if I run the program using the `mpiexec` command. Otherwise, if I run my program with only one node without using `mpiexec`, my program does everything ending up with an error.

**Comment by ddemidov** [comment]
How do you partition your problem? Can you provide a minimal reproducible example?

**Comment by ddemidov** [comment]
I think you just need to replace these two lines:
```cpp
        Solver solve(world_amgcl, std::tie(chunk, ptr, col, val), prm, bprm);
        std::unique_ptr<Solver> amg_solver = std::make_unique<Solver>(solve);
```
with
```cpp
        auto solve = std::make_unique<Solver>(world_amgcl, std::tie(chunk, ptr, col, val), prm, bprm);
```
and use the `solve` below as:
```cpp
        std::tie(iters, resid) = (*solve)(*f, *x);
```

The problem with your code is you are trying to construct two solver objects: first, by calling normal constructor, and second, by calling a copy constructor (with `make_unique`). I never intended to make solver instances copyable, so I should probably make this case into a compile error.

**Comment by ddemidov** [comment]
With bf38226 you should get a compile error instead of runtime error when trying to use a copy constructor with a `make_solver` instance.

**Comment by dboaretti** [comment]
I solved my problem, the update of the global solution among the nodes was missing in my code and the lower and higher values of the corners were not correct (there was a -1 missing). Now the solver works with every configuration. 
Thank you again for your help.
